### DISCUSSION TITLE
PaoPaoGateway socks5 模式下 fast_node 检测 看起来也生效了

### DISCUSSION THREAD

**Discussion by kbrx93** [kkkgo/PaoPaoGateWay#19] [问和答] [answered]
背景：PaoPaoGateway 采用 socks5 模式时，当 fast_node 启用时，目前网关还是会先去检测 204 通不通，不通的话不会转发流量到 s5 上。

```ini
# yaml|suburl|ovpn mode setting
# fast_node=check/yes/no
fast_node=yes
```

当然可以直接关掉 fast_node。这里主要想讨论下 socks5 模式下这个检测是不是可以直接跳过。

我本地的需求是因为代理规则比较复杂，所以启用了 socks5。并且还有个附带好处：当隧道全挂了，直接开一个本地的 socks5 接上，就可以在 FakeIP 保留的情况最大程度避免影响，不会出现因为 "DNS污染" 导致的短暂不可联网。

但是在测试的时候发现，网关接上 本地的 socks5 后，流量一直没有转发过去，看 log 发现一直在 检测 node 失败。
实际上按我个人的理解，socks5 本质上只需要检测 socks5 是否国内联通即可，更多的控制可以放在网关之外了，网关只做网关干的活。

> 之前也是看注释😂，以为 socks5 模式这个选项不会生效，就没管了。

最后还是非常感谢作者，这个作品目前用下来感觉挺好，比起自己之前手动搭的一套 FakeIP + OSPF 简单不少 👍🏻

**Comment by kkkgo** [comment, marked as answer]
按之前的设计来说是跳过的，但其实fast_node不仅是测速的功能，如果网关不能通过测试url，说明可能有哪里存在问题，比如clash程序崩溃，这时候fast_node就相当于一个看门狗，会自动重载所有服务，直到网关通过测试为止，因此他在socks5模式下其实也是有用的，只是设置为yes和check没啥区别。  
另外如果代理规则只是选择不同的域名代不代理的话，可以由PaoPaoDNS来完成，除非是你有不同域名走不同节点这样的复杂需求，那确实可以使用socks模式。
此外yaml模式也可以编写代理规则，如果能满足你的需求的话。还有一些第三方的订阅转换服务/程序，可能也可以加入规则来使用suburl模式。
最后谢谢你的使用分享👏

**Reply by kkkgo** [reply]
因为 "DNS污染" 导致的短暂不可联网的问题的另一种解决方式，可以参考“可选：生成前置嗅探的ISO” 一节。

**Reply by kkkgo** [reply]
设置为yes和check其实还是有点区别的，设置为check不会测速。测速失败到阈值会杀死进程并终止应用网关，而check不会杀死进程，仅重载所有配置并关闭所有旧连接。

**Reply by kkkgo** [reply]
另外补充一下：
- FAKE IP在网关是有缓存的，除非重启网关或者修改ppgw.ini的fakeip段设置，FAKE IP的映射一般不会变，只要你不重启网关，一般不用担心这个问题，无论是切换模式或者切换节点都不会导致FAKEIP映射产生变化；  
- 网关的角色其实相当于一条区别于正常出口的线路，设计的用途是把有需要的流量静态路由过来，也就是FAKEIP段，对大多数用途来说，他相当于一个只会出国的网关，因此socks5 也就不会去检测国内是否联通，你可以根据自己的实际需求，修改默认的测速URL。

**Reply by kbrx93** [reply]
嗯，如果是当作 watchdog 来用的话也能理解。😂

目前是用 PaoPaoDNS + PaoPao网关的方式，用 socks5 方式除了 clash 用不太顺手之外，还有个问题，就是之前用 FakeIP ，发现隧道与网关在同一节点不是一个特别好的选择。相比于网关，隧道本身的稳定性要差一些，就算加了负载也可能会有一些场外因素或者软件因素影响，这个时候就不得不重启网关，导致 FakeIP 重新映射。

前置嗅探的方式先用一段时候看看效果，也是个简化的思路。🙌

**Reply by kbrx93** [reply]
PS：ISO + 配置下发真是个好主意，很方便，同时还不失灵活性。👍🏻  不过也提个小小的建议，其实 定制 ISO 完全可以用 Github Action 来做，感觉这样子更方便一些，只需要 fork 一下仓库就好了

**Reply by kkkgo** [reply]
Github Action可以完成这个目的，但Github Action生成需要下载后解压，对于反复调试配置生成ISO不是很方便，docker的话只需要pull一次就可以离线在本地无限次生成ISO在本地，直接复制就可以马上测试，而且修改内嵌配置在本地都能直接记事本修改，替换核心也很方便，复制粘贴就可以，如果网页操作就可能麻烦点，更新版本的话也只需要直接pull就可以，fork还需要手动同步。Windows的话其实可以直接安装Docker Desktop，命令使用也是一样的，不需要专门到Linux上生成ISO，Release上的ISO确实是由Github Actions生成的，不过我自己测试也懒得从网页下载了，都是直接docker pull一下就到本地直接测试。前置嗅探其实我自己也在用，因为经常折腾各种测试，干脆就嗅探了😄

**Reply by kbrx93** [reply]
嗯，调试的话确实本地比较方便，主要是觉得 Action 对于想稍微试下的人会更友好点，可能就只是想改下网关信息之类。最主要还是想这个项目能给更多人用起来，个人感觉还是很好用的👏 。

**Comment by hpwater** [comment]
生效是生效，但是解析不出dns地址为7.*.*.*的模式

**Reply by kkkgo** [reply]
参考https://github.com/kkkgo/PaoPaoGateWay/discussions/23 配置，如有问题重新另开贴。
